File: quasar-backend/src/controllers/gamificationController.js
const UserStats = require('../models/UserStats');
const { LEVELS, ALL_BADGES } = require('../services/gamificationService');
const logger = require('../utils/logger');

/**
 * GET /api/users/:userId/stats
 * Full XP / level / streak / badge profile for a user.
 * Requester must be the owner of the stats.
 */
async function getUserStats(req, res) {
  try {
    const { userId } = req.params;

    if (req.user.id !== userId) {
      return res.status(403).json({ success: false, message: 'Forbidden', data: null });
    }

    let stats = await UserStats.findOne({ userId }).lean();

    // Return sensible defaults if no stats doc has been created yet
    if (!stats) {
      stats = {
        xp: 0,
        level: 1,
        currentStreak: 0,
        longestStreak: 0,
        lastPracticeDate: null,
        totalSessions: 0,
        badges: [],
        domainsPlayed: [],
        personasUsed: [],
        jdsParsed: 0,
        resumeComparesRun: 0,
      };
    }

    const xp    = stats.xp    || 0;
    const level = stats.level || 1;
    // LEVELS is 0-indexed; LEVELS[level] is the XP threshold for (level + 1)
    const nextThreshold = level < LEVELS.length ? LEVELS[level] : null;
    const xpToNextLevel = nextThreshold !== null ? nextThreshold - xp : null;

    return res.json({
      success: true,
      message: 'Stats retrieved',
      data: {
        xp,
        level,
        xpToNextLevel,
        currentStreak:    stats.currentStreak    || 0,
        longestStreak:    stats.longestStreak    || 0,
        lastPracticeDate: stats.lastPracticeDate || null,
        totalSessions:    stats.totalSessions    || 0,
        badges:           stats.badges           || [],
        domainsPlayed:    stats.domainsPlayed    || [],
        personasUsed:     stats.personasUsed     || [],
        jdsParsed:        stats.jdsParsed        || 0,
        resumeComparesRun: stats.resumeComparesRun || 0,
      },
    });
  } catch (err) {
    logger.error('Get user stats error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get stats', data: null });
  }
}

/**
 * GET /api/users/:userId/badges
 * Returns all 25 badges with unlocked/locked status.
 */
async function getUserBadges(req, res) {
  try {
    const { userId } = req.params;

    if (req.user.id !== userId) {
      return res.status(403).json({ success: false, message: 'Forbidden', data: null });
    }

    const stats = await UserStats.findOne({ userId }).select('badges').lean();
    const earned = stats?.badges || [];

    // Build full catalogue with unlock status
    const catalogue = Object.entries(ALL_BADGES).map(([id, meta]) => {
      const match = earned.find(b => b.id === id);
      return {
        id,
        ...meta,
        unlocked:   !!match,
        unlockedAt: match?.unlockedAt || null,
      };
    });

    return res.json({
      success: true,
      message: 'Badges retrieved',
      data: {
        earned: earned.length,
        total:  catalogue.length,
        badges: catalogue,
      },
    });
  } catch (err) {
    logger.error('Get user badges error', { err: err.message });
    return res.status(500).json({ success: false, message: 'Failed to get badges', data: null });
  }
}

module.exports = { getUserStats, getUserBadges };

File: backend/src/main/java/com/teamzemo/quasar/controller/GamificationController.java
package com.teamzemo.quasar.controller;

import com.teamzemo.quasar.dto.ApiResponse;
import com.teamzemo.quasar.model.Badge;
import com.teamzemo.quasar.model.UserStats;
import com.teamzemo.quasar.repository.UserStatsRepository;
import com.teamzemo.quasar.security.UserPrincipal;
import com.teamzemo.quasar.service.GamificationService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api")
public class GamificationController {

    private static final Logger log = LoggerFactory.getLogger(GamificationController.class);

    private final UserStatsRepository userStatsRepository;

    public GamificationController(UserStatsRepository userStatsRepository) {
        this.userStatsRepository = userStatsRepository;
    }

    /** GET /api/users/:userId/stats */
    @GetMapping("/users/{userId}/stats")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getUserStats(
            @PathVariable String userId,
            @AuthenticationPrincipal UserPrincipal principal) {
        try {
            if (!principal.getId().equals(userId)) {
                return ResponseEntity.status(403).<ApiResponse<Map<String, Object>>>body(ApiResponse.fail("Forbidden"));
            }

            UserStats stats = userStatsRepository.findByUserId(userId).orElse(null);

            int xp = 0, level = 1, currentStreak = 0, longestStreak = 0,
                totalSessions = 0, jdsParsed = 0, resumeComparesRun = 0;
            String lastPracticeDate = null;
            List<Badge> badges  = List.of();
            List<String> domainsPlayed = List.of();
            List<String> personasUsed  = List.of();

            if (stats != null) {
                xp                = stats.getXp()               != null ? stats.getXp()               : 0;
                level             = stats.getLevel()            != null ? stats.getLevel()            : 1;
                currentStreak     = stats.getCurrentStreak()    != null ? stats.getCurrentStreak()    : 0;
                longestStreak     = stats.getLongestStreak()    != null ? stats.getLongestStreak()    : 0;
                lastPracticeDate  = stats.getLastPracticeDate();
                totalSessions     = stats.getTotalSessions()    != null ? stats.getTotalSessions()    : 0;
                jdsParsed         = stats.getJdsParsed()        != null ? stats.getJdsParsed()        : 0;
                resumeComparesRun = stats.getResumeComparesRun()!= null ? stats.getResumeComparesRun(): 0;
                badges            = stats.getBadges()           != null ? stats.getBadges()           : List.of();
                domainsPlayed     = stats.getDomainsPlayed()    != null ? stats.getDomainsPlayed()    : List.of();
                personasUsed      = stats.getPersonasUsed()     != null ? stats.getPersonasUsed()     : List.of();
            }

            int[] LEVELS = GamificationService.LEVELS;
            Integer xpToNextLevel = (level < LEVELS.length) ? LEVELS[level] - xp : null;

            Map<String, Object> data = new LinkedHashMap<>();
            data.put("xp",               xp);
            data.put("level",            level);
            data.put("xpToNextLevel",    xpToNextLevel);
            data.put("currentStreak",    currentStreak);
            data.put("longestStreak",    longestStreak);
            data.put("lastPracticeDate", lastPracticeDate);
            data.put("totalSessions",    totalSessions);
            data.put("badges",           badges);
            data.put("domainsPlayed",    domainsPlayed);
            data.put("personasUsed",     personasUsed);
            data.put("jdsParsed",        jdsParsed);
            data.put("resumeComparesRun",resumeComparesRun);

            return ResponseEntity.ok(ApiResponse.ok("Stats retrieved", data));
        } catch (Exception e) {
            log.error("Get user stats error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get stats"));
        }
    }

    /** GET /api/users/:userId/badges */
    @GetMapping("/users/{userId}/badges")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getUserBadges(
            @PathVariable String userId,
            @AuthenticationPrincipal UserPrincipal principal) {
        try {
            if (!principal.getId().equals(userId)) {
                return ResponseEntity.status(403).<ApiResponse<Map<String, Object>>>body(ApiResponse.fail("Forbidden"));
            }

            UserStats stats = userStatsRepository.findByUserId(userId).orElse(null);
            List<Badge> earned = stats != null && stats.getBadges() != null ? stats.getBadges() : List.of();

            Set<String> earnedIds = earned.stream().map(Badge::getId).collect(Collectors.toSet());

            List<Map<String, Object>> catalogue =
                    GamificationService.ALL_BADGES.entrySet().stream().map(e -> {
                        String id   = e.getKey();
                        Map<String, String> meta = e.getValue();
                        Badge match  = earned.stream()
                                .filter(b -> id.equals(b.getId()))
                                .findFirst().orElse(null);
                        Map<String, Object> item = new LinkedHashMap<>();
                        item.put("id",          id);
                        item.put("name",        meta.get("name"));
                        item.put("description", meta.get("description"));
                        item.put("icon",        meta.get("icon"));
                        item.put("unlocked",    earnedIds.contains(id));
                        item.put("unlockedAt",  match != null ? match.getUnlockedAt() : null);
                        return item;
                    }).collect(Collectors.toList());

            return ResponseEntity.ok(ApiResponse.ok("Badges retrieved", Map.of(
                    "earned",  earned.size(),
                    "total",   catalogue.size(),
                    "badges",  catalogue
            )));
        } catch (Exception e) {
            log.error("Get user badges error", e);
            return ResponseEntity.internalServerError().body(ApiResponse.fail("Failed to get badges"));
        }
    }
}
